src/rgw/rgw_rest_s3.cc

```diff
@@ -713,45 +735,83 @@ void RGWListBucket_ObjStore_S3::send_versioned_response()
   s->formatter->dump_string("VersionIdMarker", marker.instance);
   if (is_truncated && !next_marker.empty()) {
     s->formatter->dump_string("NextKeyMarker", next_marker.name);
-    s->formatter->dump_string("NextVersionIdMarker", next_marker.instance);
+    if (next_marker.instance.empty())
+      s->formatter->dump_string("NextVersionIdMarker", "null");  
+    else
+      s->formatter->dump_string("NextVersionIdMarker", next_marker.instance);
   }
   s->formatter->dump_int("MaxKeys", max);
   if (!delimiter.empty())
     s->formatter->dump_string("Delimiter", delimiter);
 
   s->formatter->dump_string("IsTruncated", (max && is_truncated ? "true"
-					    : "false"));
+              : "false"));
+
 
-  bool encode_key = false;
-  if (strcasecmp(encoding_type.c_str(), "url") == 0) {
-    s->formatter->dump_string("EncodingType", "url");
-    encode_key = true;
-  }
 
   if (op_ret >= 0) {
     if (objs_container) {
       s->formatter->open_array_section("Entries");
     }
 
```

Comment:
Have you accidentally removed code that has to go in for Objects here?

Reply:
Isn't it better to add encode_key definition in v1 and v2 seperately because encode_key is used inside a loop that can't be made common since some elements are unique for v1 and v2?

Reply:
then you should have an open_section and a corresponding close section call in v1 and v2.

Reply:
Added.